cristin-organization/src/main/java/no/unit/nva/cristin/organization/common/client/v20230526/OrganizationEnricher.java

```diff
@@ -0,0 +1,79 @@
+package no.unit.nva.cristin.organization.common.client.v20230526;
+
+import static no.unit.nva.cristin.model.JsonPropertyNames.DEPTH;
+import static no.unit.nva.cristin.model.JsonPropertyNames.IDENTIFIER;
+import static nva.commons.core.attempt.Try.attempt;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import no.unit.nva.client.FetchApiClient;
+import no.unit.nva.model.Organization;
+import no.unit.nva.utils.UriUtils;
+import nva.commons.apigateway.exceptions.ApiGatewayException;
+import nva.commons.apigateway.exceptions.NotFoundException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class OrganizationEnricher {
+
+    private static final Logger logger = LoggerFactory.getLogger(OrganizationEnricher.class);
+
+    public static final String COULD_NOT_BE_FOUND_IN_UPSTREAM =
+        "Organization from search result could not be found in upstream: {}";
+
+    private final List<Organization> inputOrganizations;
+    private final Map<String, String> queryParams;
+    private final FetchApiClient<Map<String, String>, Organization> fetchClient;
+    private final List<Organization> enrichedOrganizations;
+
+    public OrganizationEnricher(List<Organization> inputOrganizations,
+                                Map<String, String> queryParams,
+                                FetchApiClient<Map<String, String>, Organization> fetchClient) {
+        this.inputOrganizations = inputOrganizations;
+        this.queryParams = queryParams;
+        this.fetchClient = fetchClient;
+        enrichedOrganizations = new ArrayList<>();
+    }
+
+    public OrganizationEnricher enrich() throws ApiGatewayException {
+        for (Organization organization : inputOrganizations) {
+            var identifier = extractIdentifier(organization);
+            var enrichParams = extractParams(identifier);
+            try {
+                var enriched = fetchClient.executeFetch(enrichParams);
+                enrichedOrganizations.add(enriched);
+            } catch (NotFoundException notFoundException) {
+                logger.warn(COULD_NOT_BE_FOUND_IN_UPSTREAM, extractOrgId(organization));
+            }
+        }
+
+        return this;
+    }
+
+    public List<Organization> getResult() {
+        return enrichedOrganizations;
+    }
```

Comment:
Since you only ever call `getResult()` after calling `enrich()`, maybe something like:

```

    public List<Organization> enrich() {
        return inputOrganizations.stream()
                .map(this::fetch)
                .filter(Optional::isPresent)
                .map(Optional::get)
                .collect(Collectors.toList());
    }

    private Optional<Organization> fetch(Organization organization) {
        var attempt = attempt(() -> fetchClient.executeFetch(extractParams(organization)));
        return attempt.isSuccess() ? Optional.of(attempt.get()) : handleException(attempt, organization);
    }

    private Optional<Organization> handleException(Try<Organization> attempt, Organization organization) {
        if (attempt.getException() instanceof NotFoundException) {
            logger.warn(COULD_NOT_BE_FOUND_IN_UPSTREAM, extractOrgId(organization));
            return Optional.empty();
        }
        throw new RuntimeException(attempt.getException());
    }
```

Reply:
I did it this way in case i wanted to do more mutations on the state before retrieving the result, as is good practice in OOP. But if you think it is unnecessary, i can just return the object directly without having it in a separate method. If we wanted to add more logic later, having the result in a separate method could be a good thing...

Reply:
This will swallow exceptions related to upstream (for instance 502 Bad Gateway) and instead give a generic 500 Server Error even when there is no server error. That is why i had to do it without using functional style.

I can refactor if you think this is not a problem, will await your feecback

Reply:
Are you thinking of a case where you do e.g. 

```
            var organizationEnricher = new OrganizationEnricher(organizations, params, fetchClient);
            var organizations = organizationEnricher.enrich()
                 .translateLabels() // may be optionally and independently performed of "enrich()"
                 .addSubunitCount() // may be optionally and independently performed of "enrich()"
                 .getResult();
```

Reply:
Yes, something like that.